doc/developer/design/20230110_window_functions.md

```diff
@@ -0,0 +1,255 @@
+```
+
+(And this is just the MIR, but we would work with the LIR.)
+
+Also, we will want other transforms for window functions, most notably, the ROW_NUMBER-to-TopK transform. For this, we need to canonicalize scalar expressions, which I think we usually do in MIR. This means that this transform should happen on MIR, meaning that we would need an MIR representation of window functions.
+
+### Extended discussion on each of the IRs
+
+**HIR**
+
+In HIR, the windowing is currently in the scalar expressions (option 2. from above), but it’s possible to change this.
+
+1. *Dedicated `HirRelationExpr` variant:*
+    - There is a precedent for a similar situation: HIR has aggregation expressions, which (similarly to window expressions) have the property that they are in a scalar expression position, but their value is actually calculated by a dedicated `HirRelationExpr` variant (`Reduce`), and then there is just a column reference in `HirScalarExpr`.
+2. *Hiding in `HirScalarExpr`:*
+    - HIR wants to be close to the SQL syntax, and windowing appears in scalar position in SQL.
+    - It’s already implemented this way, so if there is no strong argument for 1. or 3., then I’d like to just leave it as it is.
+3. *Reusing `Reduce`*
+
+**MIR**
+
+We need an MIR representation for two things:
+
+- To get the window function expressions to rendering, where we’ll apply the prefix sum. (Alternatively, we could recover windowing from the patterns that get created when the current HIR lowering compiles away the windowing.)
+- To have optimizer transforms for some important special cases of window functions, e.g., for TopK patterns. (Alternatively, we could apply these transforms in the HIR-to-MIR lowering.)
+1. *Create a dedicated enum variant in `MirRelationExpr`:*
+    - I think this is better than 2., because Map (and MirScalarExprs in general) should have the semantics that they can be evaluated by just looking at one input element, while a window function needs to look at large parts of a window partition. If we were to put window functions into scalar expressions, then we would need to check lots of existing code that is processing MirScalarExprs that they are not getting unpleasantly surprised by window functions.
+    - Compared to 3., it might be easier to skip windowing in many transforms. This is both good and bad:
+        - We can get a first version done more quickly. (And then potentially add optimizations later.)
+        - But we might leave some easy optimization opportunities on the table, which would come from already-existing transform code for `Reduce`.
+   - A new `MirRelationExpr` variant would mean we have to modify about 12-14 transforms (`LetRec` is pattern-matched in 12 files in the `transform` crate, `TopK` 14 times. See also [the recent LetRec addition](https://github.com/MaterializeInc/materialize/commit/9ac8e060d82487752ba28c42f7b146ff9f730ca3) for an example of how it looks when we add a new `MirRelationExpr` variant.)
```

Comment:
For clarity, `LetRec` is disabled in all but a few transforms because we don't know that it is correct without a more thorough audit of the transforms. We have non-exhaustive matches (e.g. predicate pushdown) where I suspect it is not correct, and .. we'll have to do the same sort of checking for a `WindowFunction` variant).

Reply:
I changed the text to clarify this.